ddl/ddl_api.go

```diff
@@ -3519,32 +3526,111 @@ func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, keyType ast.Inde
+			if err != nil {
+				return errors.Trace(err)
+			}
+			expr, err := expression.RewriteSimpleExprWithTableInfo(ctx, tblInfo, idxPart.Expr)
+			if err != nil {
+				// TODO: refine the error message
+				return err
+			}
+			if _, ok := expr.(*expression.Column); ok {
+				return ErrFunctionalIndexOnField
+			}
+
+			colInfo := &model.ColumnInfo{
+				Name:                idxPart.Column.Name,
+				GeneratedExprString: sb.String(),
+				GeneratedStored:     false,
+				Version:             model.CurrLatestColumnInfoVersion,
+				Dependences:         make(map[string]struct{}),
+				Hidden:              true,
+				FieldType:           *expr.GetType(),
+			}
+			for _, colName := range findColumnNamesInExpr(idxPart.Expr) {
+				colInfo.Dependences[colName.Name.O] = struct{}{}
+			}
+			if err = checkDependedColExist(colInfo.Dependences, t.Cols()); err != nil {
+				return errors.Trace(err)
+			}
+			idxPart.Expr = nil
+			hiddenCols = append(hiddenCols, colInfo)
+		}
+	}
+
 	// Check before the job is put to the queue.
 	// This check is redundant, but useful. If DDL check fail before the job is put
 	// to job queue, the fail path logic is super fast.
 	// After DDL job is put to the queue, and if the check fail, TiDB will run the DDL cancel logic.
 	// The recover step causes DDL wait a few seconds, makes the unit test painfully slow.
-	_, err = buildIndexColumns(tblInfo.Columns, idxColNames)
+	_, err = buildIndexColumns(append(tblInfo.Columns, hiddenCols...), indexPartSpecifications)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	if unique && tblInfo.GetPartitionInfo() != nil {
-		if err := checkPartitionKeysConstraint(tblInfo.GetPartitionInfo(), idxColNames, tblInfo, false); err != nil {
+		if err := checkPartitionKeysConstraint(tblInfo.GetPartitionInfo(), indexPartSpecifications, tblInfo, false); err != nil {
 			return err
 		}
 	}
 	// May be truncate comment here, when index comment too long and sql_mode is't strict.
 	if _, err = validateCommentLength(ctx.GetSessionVars(), indexName.String(), indexOption); err != nil {
 		return errors.Trace(err)
 	}
-
+	if len(hiddenCols) > 0 {
+		// Check hidden column
+		if err = checkAddColumnTooManyColumns(len(t.Cols()) + len(hiddenCols)); err != nil {
+			return errors.Trace(err)
+		}
+		// Check whether the hidden columns have existed.
+		for _, colInfo := range hiddenCols {
+			// Check whether the hidden columns have existed.
+			col := table.FindCol(t.Cols(), colInfo.Name.String())
+			if col != nil {
+				// TODO: use expression index related error
+				return infoschema.ErrColumnExists.GenWithStackByArgs(colInfo.Name.String())
+			}
+			if err = checkAutoIncrementRef("", colInfo.Dependences, tblInfo); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for _, indexPart := range indexPartSpecifications {
+			if indexPart.Expr != nil {
+				// TODO: refine the error message
+				if err := checkIllegalFn4GeneratedColumn("", indexPart.Expr); err != nil {
```

Comment:
Why do we need to do it again? It seems that we called "checkIllegalFn4GeneratedColumn" in line3541.

Reply:
Forget to remove it.